README.md
Overview
========

This package provides a redis cache manager for Zope. It also makes
available a decorator from dogpile.cache, which can be used to cache
specific functions or methods.

Installation
============

Add `collective.rediscache` to the buildout eggs. The package supports
configuration using the Zope product configuration mechanism:

::

        <product-config collective.rediscache>
            redis_url redis://127.0.0.1:6379
            expiration_time 300
        </product-config>

You can add this snippet to the zope-conf-additional entry in your buildout,
or directly into your instance's zope.conf file, if not using buildout.

`redis_url` is the URL for the Redis service, and `expiration_time` is the
cache entry expiration time in seconds.

Cache Manager
=============

The cache manager is compatible with any cache that uses OFS.Cache, like the
standard cache managers from Plone. To use it, it's necessary to use the ZMI's
"add" menu. Usually, it should be added to the root of the Plone site. Once
you add a `Redis Cache Manager`, you can associate it with any cacheable Zope
object, using the `cache` tab in the object's ZMI view. Associated objects
will use the cache to set and get values after that.

The cache has a single editable property, `request variables`. This property
accepts a request variable name on each line. These request variables will be
used to generate the key under which this object will be stored in the cache.
By default, the only variable used is `AUTHENTICATED_USER`. This means that
a value will be stored for the same object for each user that calls it.

To invalidate cache entries, use the `invalidate` button from the object's
cache tab in the ZMI. Note that this will invalidate all stored values for
that object, even if they come from several views.

Cache Decorator
===============

The package defines a dogpile.cache region that can be used to decorate
functions and methods. This will cache the return value of the function using
a key derived from its arguments. Example:

::
  
    from collective.rediscache.cachemanager import redis_cache


    @redis_cache.cache_on_arguments()
    def some_expensive_call(x, y):
        return very_slow_function(x) + ultra_slow_function(y)

`cache_on_arguments` can include the keyword parameter `expiration_time`,
given in seconds. This expiration time  will override the general cache
expiration time configuration. The decorated function can be called normally,
but the return value will be stored in the cache and pulled from there until
it expires.

The decorator also adds a couple of useful attributes to the function. The
`invalidate` attribute invalidates the return value generated with the
arguments used:

::

    some_expensive_call.invalidate(3, 4)

The `set` attribute sets a value without calling the functions, and assigns
it to the key generated by the regular arguments.

::

    some_expensive_call.set(new_value, 3, 4)

The `original` attribute is used to call the function directly, without using
the cache:

::

    some_expensive_call.original(3, 4)

For more information on the decorator, consult the dogpile.cache documentation.


CHANGES.rst
Changelog
=========

0.2.0 (unreleased)
------------------

- Switch logging statements to use debug instead of info.
  [enfold-josh]


0.1.0 (2018-08-08)
------------------

- Initial release.


collective/rediscache/cachemanager.py
from logging import getLogger
import time

from dogpile.cache import make_region
from dogpile.cache.api import NO_VALUE
from dogpile.cache.proxy import ProxyBackend

from AccessControl.class_init import InitializeClass
from AccessControl.Permissions import view_management_screens
from AccessControl.SecurityInfo import ClassSecurityInfo
from App.config import getConfiguration
from App.special_dtml import DTMLFile
from OFS.Cache import Cache
from OFS.Cache import CacheManager
from OFS.SimpleItem import SimpleItem


logger = getLogger('collective.rediscache')

caches = {}
DEFAULT_URL = 'redis://127.0.0.1:6379'
KEY_PREFIX = 'collective.rediscache:'


class LoggingProxy(ProxyBackend):

    def get(self, key):
        value = self.proxied.get(key)
        if 'ZCache:' not in key:
            result = "HIT"
            if value is NO_VALUE:
                result = "MISS"
            logger.debug("Cache {} for key {}".format(result, key))
        return value

    def set(self, key, value):
        if 'ZCache:' not in key:
            logger.debug("Setting value for key {}".format(key))
        return self.proxied.set(key, value)

redis_cache = make_region(
    name='collective.rediscache',
    key_mangler=lambda key: KEY_PREFIX + key,
)


def init_cache():
    zope_conf = getConfiguration()
    product_conf = getattr(zope_conf, 'product_config', {})
    cache_conf = product_conf.get('collective.rediscache', {})
    redis_url = cache_conf.get('redis_url', DEFAULT_URL)
    expiration_time = cache_conf.get('expiration_time', 300)
    cache = redis_cache.configure(
        'dogpile.cache.redis',
        expiration_time=int(expiration_time),
        arguments={
            'url': redis_url,
            'distributed_lock': True,
        },
        wrap=[LoggingProxy],
    )
    return cache


class RedisCache(Cache):
    max_age = 0

    def __init__(self):
        self.cache = redis_cache
        self.next_cleanup = 0

    def initSettings(self, kw):
        self.__dict__.update(kw)

    def get_key(self, ob, view_name, keywords):
        if not view_name:
            view_name = '__default__'
        path = ob.absolute_url_path()
        req = getattr(ob, 'REQUEST', None)
        req_vals = '|'
        for key in self.request_vars:
            if req is None:
                val = ''
            else:
                val = req.get(key, '')
            req_vals += "%s %s " % ((str(key), str(val)))
        keyword_vals = ''
        if keywords:
            keyword_vals = '|'
            for key, val in keywords.items():
                keyword_vals += "%s %s " % ((str(key), str(val)))
        return 'ZCache:' + path + ':' + view_name + req_vals + keyword_vals

    def ZCache_invalidate(self, ob):
        '''
        Invalidates the cache entries that apply to ob.
        '''
        path = ob.absolute_url_path()
        regex = "%s%s*" % (KEY_PREFIX, path)
        # need to get the client because scanning for regex
        client = self.cache.actual_backend.client
        entries = client.scan_iter(match=regex)
        for entry in entries:
            # delete goes through key mangler, so get rid of prefix
            self.cache.delete(entry[len(KEY_PREFIX):])
        logger.info("Invalidated all cache entries for {}".format(path))

    def ZCache_get(self, ob, view_name='', keywords=None,
                   mtime_func=None, default=None):
        '''
        Gets a cache entry or returns default.
        '''
        result = "HIT"
        path = ob.absolute_url_path()
        key = self.get_key(ob, view_name, keywords)
        value = self.cache.get(key)
        if value is NO_VALUE:
            value = default
            result = 'MISS'
        logger.debug("Cache {} for {} (key={})".format(result, path, key))
        return value

    def ZCache_set(self, ob, data, view_name='', keywords=None,
                   mtime_func=None):
        '''
        Sets a cache entry.
        '''
        path = ob.absolute_url_path()
        key = self.get_key(ob, view_name, keywords)
        self.cache.set(key, data)
        logger.debug("Setting value for {} (key={})".format(path, key))


class RedisCacheManager(CacheManager, SimpleItem):
    """Manage a RedisCache, which stores rendered data in redis.

    This is intended to be used as a low-level cache for
    expensive Python code, not for objects published
    under their own URLs such as web pages.

    RedisCacheManager *can* be used to cache complete publishable
    pages, such as DTMLMethods/Documents and Page Templates,
    but this is not advised: such objects typically do not attempt
    to cache important out-of-band data such as 3xx HTTP responses,
    and the client would get an erroneous 200 response.

    Such objects should instead be cached with an
    AcceleratedHTTPCacheManager and/or downstream
    caching.
    """

    security = ClassSecurityInfo()
    security.setPermissionDefault('Change cache managers', ('Manager', ))

    manage_options = (
        {'label': 'Properties', 'action': 'manage_main'},
    ) + CacheManager.manage_options + SimpleItem.manage_options

    meta_type = 'Redis Cache Manager'

    def __init__(self, ob_id):
        self.id = ob_id
        self.title = ''
        self._settings = {
            'request_vars': ('AUTHENTICATED_USER', ),
            }
        self._resetCacheId()

    def getId(self):
        ' '
        return self.id

    security.declarePrivate('_remove_data')
    def _remove_data(self):
        caches.pop(self.__cacheid, None)

    security.declarePrivate('_resetCacheId')
    def _resetCacheId(self):
        self.__cacheid = '%s_%f' % (id(self), time.time())

    ZCacheManager_getCache__roles__ = ()
    def ZCacheManager_getCache(self):
        cacheid = self.__cacheid
        try:
            return caches[cacheid]
        except KeyError:
            cache = RedisCache()
            cache.initSettings(self._settings)
            caches[cacheid] = cache
            return cache

    security.declareProtected(view_management_screens, 'getSettings')
    def getSettings(self):
        'Returns the current cache settings.'
        res = self._settings.copy()
        return res

    security.declareProtected('Change cache managers', 'manage_editProps')
    def manage_editProps(self, title, settings=None, REQUEST=None):
        'Changes the cache settings.'
        if settings is None:
            settings = REQUEST
        self.title = str(title)
        request_vars = list(settings['request_vars'])
        request_vars.sort()
        self._settings = {
            'request_vars': tuple(request_vars),
            }
        cache = self.ZCacheManager_getCache()
        cache.initSettings(self._settings)
        if REQUEST is not None:
            return self.manage_main(
                self, REQUEST, manage_tabs_message='Properties changed.')

    security.declareProtected(view_management_screens, 'manage_main')
    manage_main = DTMLFile('dtml/propsRCM', globals())


InitializeClass(RedisCacheManager)


manage_addRedisCacheManagerForm = DTMLFile('dtml/addRCM', globals())


def manage_addRedisCacheManager(self, id, REQUEST=None):
    'Adds a Redis cache manager to the folder.'
    self._setObject(id, RedisCacheManager(id))
    if REQUEST is not None:
        return self.manage_main(self, REQUEST)
